# MUHAMMED SHAHIL T

# E-MAIL SPAM DETECTION

### Importing libraries

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import warnings
warnings.filterwarnings('ignore')

### Importing the dataset

In [22]:
df=pd.read_csv(r"C:\Users\shahi\Documents\Downloads\spam.csv",encoding=('ISO-8859-1'),low_memory=False)

In [23]:
df

v1                                                 v2 Unnamed: 2  \
0      ham  Go until jurong point, crazy.. Available only ...        NaN   
1      ham                      Ok lar... Joking wif u oni...        NaN   
2     spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3      ham  U dun say so early hor... U c already then say...        NaN   
4      ham  Nah I don't think he goes to usf, he lives aro...        NaN   
...    ...                                                ...        ...   
5567  spam  This is the 2nd time we have tried 2 contact u...        NaN   
5568   ham              Will Ì_ b going to esplanade fr home?        NaN   
5569   ham  Pity, * was in mood for that. So...any other s...        NaN   
5570   ham  The guy did some bitching but I acted like i'd...        NaN   
5571   ham                         Rofl. Its true to its name        NaN   

     Unnamed: 3 Unnamed: 4  
0           NaN        NaN  
1           NaN        NaN  
2           NaN        NaN  
3           NaN        NaN  
4           NaN        NaN  
...         ...        ...  
5567        NaN        NaN  
5568        NaN        NaN  
5569        NaN        NaN  
5570        NaN        NaN  
5571        NaN        NaN  

[5572 rows x 5 columns]

# Basic understanding of the data

In [24]:
df.head()

v1                                                 v2 Unnamed: 2  \
0   ham  Go until jurong point, crazy.. Available only ...        NaN   
1   ham                      Ok lar... Joking wif u oni...        NaN   
2  spam  Free entry in 2 a wkly comp to win FA Cup fina...        NaN   
3   ham  U dun say so early hor... U c already then say...        NaN   
4   ham  Nah I don't think he goes to usf, he lives aro...        NaN   

  Unnamed: 3 Unnamed: 4  
0        NaN        NaN  
1        NaN        NaN  
2        NaN        NaN  
3        NaN        NaN  
4        NaN        NaN

In [25]:
df.dtypes

v1            object
v2            object
Unnamed: 2    object
Unnamed: 3    object
Unnamed: 4    object
dtype: object

In [26]:
df.shape

(5572, 5)

In [27]:
df.columns

Index(['v1', 'v2', 'Unnamed: 2', 'Unnamed: 3', 'Unnamed: 4'], dtype='object')

In [28]:
df.describe()

v1                      v2  \
count   5572                    5572   
unique     2                    5169   
top      ham  Sorry, I'll call later   
freq    4825                      30   

                                               Unnamed: 2  \
count                                                  50   
unique                                                 43   
top      bt not his girlfrnd... G o o d n i g h t . . .@"   
freq                                                    3   

                   Unnamed: 3 Unnamed: 4  
count                      12          6  
unique                     10          5  
top      MK17 92H. 450Ppw 16"    GNT:-)"  
freq                        2          2

In [29]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5572 entries, 0 to 5571
Data columns (total 5 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   v1          5572 non-null   object
 1   v2          5572 non-null   object
 2   Unnamed: 2  50 non-null     object
 3   Unnamed: 3  12 non-null     object
 4   Unnamed: 4  6 non-null      object
dtypes: object(5)
memory usage: 217.8+ KB


The dataset contains 5572 rows and 5 columns. 

### Renaming the columns 

In [30]:
df.rename(columns={'v1':'ham or spam'},inplace=True)


In [31]:
df.rename(columns={'v2':'mail'},inplace=True)

## Cleaning the dataset

In [32]:
df.isnull().sum()

ham or spam       0
mail              0
Unnamed: 2     5522
Unnamed: 3     5560
Unnamed: 4     5566
dtype: int64

In [33]:
df.duplicated().sum()

403

In [34]:
df=df.drop(["Unnamed: 2","Unnamed: 3","Unnamed: 4"],axis=1)

In [35]:
df=df.drop_duplicates()

In [36]:
df

ham or spam                                               mail
0            ham  Go until jurong point, crazy.. Available only ...
1            ham                      Ok lar... Joking wif u oni...
2           spam  Free entry in 2 a wkly comp to win FA Cup fina...
3            ham  U dun say so early hor... U c already then say...
4            ham  Nah I don't think he goes to usf, he lives aro...
...          ...                                                ...
5567        spam  This is the 2nd time we have tried 2 contact u...
5568         ham              Will Ì_ b going to esplanade fr home?
5569         ham  Pity, * was in mood for that. So...any other s...
5570         ham  The guy did some bitching but I acted like i'd...
5571         ham                         Rofl. Its true to its name

[5169 rows x 2 columns]

The given dataset had a number of null values and about 403 duplicate values. All the null values and duplicate values were removed. 

## Encoding the target column

In [37]:
from sklearn.preprocessing import LabelEncoder

In [38]:
lb=LabelEncoder()

In [39]:
df['ham or spam']=lb.fit_transform(df['ham or spam'])
df

ham or spam                                               mail
0               0  Go until jurong point, crazy.. Available only ...
1               0                      Ok lar... Joking wif u oni...
2               1  Free entry in 2 a wkly comp to win FA Cup fina...
3               0  U dun say so early hor... U c already then say...
4               0  Nah I don't think he goes to usf, he lives aro...
...           ...                                                ...
5567            1  This is the 2nd time we have tried 2 contact u...
5568            0              Will Ì_ b going to esplanade fr home?
5569            0  Pity, * was in mood for that. So...any other s...
5570            0  The guy did some bitching but I acted like i'd...
5571            0                         Rofl. Its true to its name

[5169 rows x 2 columns]

## Splitting the dataset

In [40]:
from sklearn.model_selection import train_test_split

In [41]:
x=df['mail']
y=df['ham or spam']

In [42]:
X_train,X_test,Y_train,Y_test=train_test_split(x,y,random_state=42,test_size=0.3)

Here, 70% of the data is used for training the data and 30% is used for testing the data.

## Vectorizing

In [43]:
from sklearn.feature_extraction.text import TfidfVectorizer

In [44]:
tfidf=TfidfVectorizer()

In [45]:
X_train_count = tfidf.fit_transform(X_train.values)
X_train_count.toarray()

array([[0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       ...,
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.],
       [0., 0., 0., ..., 0., 0., 0.]])

## Model Building

In [46]:
from sklearn.naive_bayes import MultinomialNB

In [47]:
model=MultinomialNB()

In [48]:
model.fit(X_train_count,Y_train)

MultinomialNB()

## Calculating the performance of the model

In [49]:
X_test_count = tfidf.transform(X_test)

In [50]:
model.score(X_test_count,Y_test)

0.9484203739522888

### Here we get an accuary of 94.8%

# Thank You